File: focuspoints/listComprehensions.py
# Python als Vergleich
# Vielfache von 2 von 1 bis 10
doppelte = [x * 2 for x in range(1, 11)]

# Bedingung: Nur gerade Zahlen von 1 bis 20
gerade_zahlen = [x for x in range(1, 21) if x % 2 == 0]

# Kombination mehrerer Listen
kombinationen = [(x, y) for x in [1, 2] for y in ['a', 'b']]

# Mehreren Bedingungen
zahlen_ohne_drei = [x for x in range(1, 21) if x % 3 != 0 and x % 4 != 0]

# Mit einer Funktion
quadrate = [x ** 2 for x in range(1, 11)]

# Ein kleines Beispiel, das Pythagoräische Tripel erzeugt
pythagoreische_tripel = [
    (a, b, c)
    for c in range(1, 31)
    for b in range(1, 31)
    for a in range(1, 31)
    if a * a + b * b == c * c and a < b and b < c
]

def main():
    print("Doppelte:")
    print(doppelte)

    print("\nGerade Zahlen:")
    print(gerade_zahlen)

    print("\nKombinationen:")
    print(kombinationen)

    print("\nZahlen ohne 3 und 4:")
    print(zahlen_ohne_drei)

    print("\nQuadrate:")
    print(quadrate)

    print("\nPythagoreische Tripel:")
    print(pythagoreische_tripel)

if __name__ == "__main__":
    main()

File: focuspoints/listComprehensions.hs
-- Quadrate
quadrate :: [Int]
quadrate = [x^2 | x <- [1..10]]

-- Bedingung: Nur gerade Zahlen von 1 bis 20
geradeZahlen :: [Int]
geradeZahlen = [x | x <- [1..20], even x]

-- Kombination mehrerer Listen
kombinationen :: [(Int, Char)]
kombinationen = [(x, y) | x <- [1,2], y <- ['a','b']]

-- Mehreren Bedingungen
zahlenOhneDrei :: [Int]
-- zahlenOhneDrei = [x | x <- [1..20], x `mod` 3 /= 0, x `mod` 4 /= 0]
zahlenOhneDrei = [x | x <- [1..20], (/= 0) (mod x 3), (/= 0) (mod x 4)]

-- Ein kleines Beispiel, das Pythagoräische Tripel erzeugt
pythagoreischeTripel :: [(Int, Int, Int)]
pythagoreischeTripel = [(a, b, c) |
                         c <- [1..30],
                         b <- [1..30],
                         a <- [1..30],
                         a^2 + b^2 == c^2, a < b, b < c]

main :: IO ()
main = do
    putStrLn "\nQuadrate:"
    print quadrate

    putStrLn "\nGerade Zahlen:"
    print geradeZahlen

    putStrLn "\nKombinationen:"
    print kombinationen

    putStrLn "\nZahlen ohne 3 und 4:"
    print zahlenOhneDrei

    putStrLn "\nPythagoreische Tripel:"
    print pythagoreischeTripel
